### PR TITLE
Refactor foreach binary ops tests with scalars to use OpInfo

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 [WIP] Add type promotion to a fast path for foreach APIs
* #52448 [WIP] Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* **#51058 Refactor foreach binary ops tests with scalars to use OpInfo**

-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
...
+                    torch_inplace = getattr(torch.Tensor, inplace_name, None)
+
+                    for t in tensors:
+                        torch_inplace(t, scalar)
+                except Exception:
+                    torch_inplace_exeption = True
+
+                self.assertEqual(foreach_inplace_exeption, torch_inplace_exeption)
+
+    # Test foreach binary ops with a scalar list
```
Comment (zou3519) [inline test/test_foreach.py:210]: nit: let's clarify that this is testing `binary_op(TensorList, List[Scalar])`

**Review by zou3519 — COMMENTED** [review]
Logic looks correct. I had two comments:
1. I'm not sure what is happening with the float16 testing (added an in-line comment)
2. In the old code, we had some checks for various error messages that would be raised. Those checks don't exist anymore in the OpInfo tests. Do you think it would be worth it to add some new, non-OpInfo test that tests for the various error message? I am thinking of something like:
```
def test_error_messages(self, device):
    # test subtraction with booleans raises a nice error message

    # test addition with LongTensors but with floating-point alpha raises a nice error message
    
    # ...
```
There are probably better ways to organize this. I'm OK if you want to handle adding tests for the error messages in a follow-up later to unblock progress here

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1029,6 +1029,30 @@ def __init__(self,
         self.inplace_variant = foreach_method_inplace
         self.ref = torch_ref_method
 
+class ForeachBinaryFuncInfo(OpInfo):
+    """Early version of a specialized OpInfo for foreach binary functions"""
+    def __init__(self,
+                 name,
+                 dtypes=all_types_and(torch.bool, torch.half, torch.bfloat16),
```
[the remark this answers is not in the document]
Comment (izdeby) [reply torch/testing/_internal/common_methods_invocations.py:1247]: Yeah, i agree, Its super confusing

File: test/test_foreach.py [hunk 3]
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
...
+                except Exception:
+                    foreach_inplace_exeption = True
+
+                try:
+                    # get torch inplace reference function
+                    inplace_name = op.name + "_"
+                    torch_inplace = getattr(torch.Tensor, inplace_name, None)
+
+                    for t in tensors:
+                        torch_inplace(t, scalar)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_foreach.py:205]: No need to add, just thinking out loud

File: test/test_foreach.py [hunk 4]
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
...
+    # Compare results agains reference torch functions
+    # In case of an exeption, check if torch reference function throws as well.
+    @skipCUDAIfRocm
+    @ops(foreach_binary_op_db)
+    def test_binary_ops_scalar(self, device, dtype, op):
+        scalars = [2, 2.2, True, 3 + 5j]
+
+        # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
+        dtype = torch.float32 if (self.device_type == 'cuda' and
+                                  (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_foreach.py:162]: Hmm I see, thanks for the references. Wouldn't it be better to
1. Run the foreach op in with inputs of the original dtype (float16)
2. Convert inputs to the updated dtype (float32), run the reference (PyTorch op in a for-loop), then convert outputs back to float 16
3. Compare the result of 1 and 2

This way we can still test that the foreach APIs accept float16 and bfloat16 tensors without erroring out.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1247]: It is incredibly confusing, sorry. It corresponds to the internal dispatch macro 1:1, though.

**Pull request closed without merging**